pkg/urbit/vere/io/ames.c

```diff
@@ -232,6 +301,107 @@ u3_ames_encode_lane(u3_lane lan) {
   return u3ke_jam(u3nt(c3__ipv4, u3i_words(1, &lan.pip_w), lan.por_s));
 }
 
+/* _ames_lane_from_sockaddr(): sockaddr_in to lane noun
+*/
+static u3_noun
+_ames_lane_from_sockaddr(struct sockaddr_in* add_u)
+{
+  u3_lane lan_u;
+  lan_u.por_s = ntohs(add_u->sin_port);
+  lan_u.pip_w = ntohl(add_u->sin_addr.s_addr);
+  return u3_ames_encode_lane(lan_u);
+}
+
+/* _ames_serialize_packet(): u3_panc to atom, updating the origin lane if dop_o
+**                           (retains pac_u)
+*/
+static u3_noun
+_ames_serialize_packet(u3_panc* pac_u, c3_o dop_o)
+{
+  c3_y sen_y = 2 << pac_u->hed_u.sac_y;
+  c3_y rec_y = 2 << pac_u->hed_u.rac_y;
+  c3_o nal_o = c3n;
+
+  //  update the body's lane, if desired
+  //
+  if (c3y == dop_o) {
+    //  unpack (jam [(unit lane) body])
+    //
+    u3_noun lon, bod;
+    {
+      //NOTE we checked for cue safety in _ames_recv_cb
+      u3_noun old = u3ke_cue(u3i_bytes(pac_u->bod_u.con_w, pac_u->bod_u.con_y));
+      u3x_cell(old, &lon, &bod);
+      u3k(lon); u3k(bod);
+      u3z(old);
+    }
+
+    //  only replace the lane if it was ~
+    //
+    //NOTE  this sets an opaque lane even in the "sender is galaxy" case,
+    //      but that doesn't matter: ames.hoon ignores origin in that case,
+    //      always using the appropriate galaxy lane instead.
+    //
+    if (u3_nul == lon) {
+      u3z(lon);
+      lon = u3nt(u3_nul, c3n, u3k(pac_u->ore));
+      nal_o = c3y;
+
+      c3_free(pac_u->bod_u.con_y);
+      u3_noun jam = u3ke_jam(u3nc(lon, bod));
+      pac_u->bod_u.con_w = u3r_met(3, jam);
+      pac_u->bod_u.con_y = c3_malloc(pac_u->bod_u.con_w);
+      u3r_bytes(0, pac_u->bod_u.con_w, pac_u->bod_u.con_y, jam);
+      u3z(jam);
+    }
+    else {
+      u3z(lon); u3z(bod);
+    }
+  }
+
+  //  serialize the packet
+  //
+  u3_noun pac;
+  {
+    //  start with the body
+    //
+    u3_body* bod_u = &pac_u->bod_u;
+    c3_y*    pac_y = c3_malloc(4 + sen_y + rec_y + bod_u->con_w);
+    {
+      u3_atom sen = u3i_chubs(2, bod_u->sen_d);
+      u3_atom rec = u3i_chubs(2, bod_u->rec_d);
+      u3r_bytes(0, sen_y, pac_y + 4,         sen);
+      u3r_bytes(0, rec_y, pac_y + 4 + sen_y, rec);
+      u3z(sen); u3z(rec);
+    }
+    memcpy(pac_y + 4 + sen_y + rec_y, bod_u->con_y, bod_u->con_w);
+
+    //  if we updated the origin lane, we need to update the mug too
+    //
+    if (c3y == nal_o) {
+      u3_noun bod = u3i_bytes(sen_y + rec_y + bod_u->con_w, pac_y + 4);
+      pac_u->hed_u.mug_l = u3r_mug(bod) & ((1 << 20) - 1);
+      u3z(bod);
```

Comment:
This should just be `_ca_mug_body(sen_y + rec_y + bod_u->con_w, pac_y + 4)` -- the atom allocation isn't necessary. (Also, you may need to reorder that addition, or cast the `c3_y` arguments to `c3_w`).